packages/react/src/useFetch.ts

```diff
@@ -0,0 +1,146 @@
+
+const reducer = <T>(state: FetchHookState<T>, action: FetchAction<T>): FetchHookState<T> => {
+  switch (action.type) {
+    case "fetching":
+      return { ...state, fetching: true, error: undefined };
+    case "fetched":
+      return { ...state, fetching: false, data: action.payload, error: undefined };
+    case "error":
+      return { ...state, fetching: false, error: action.payload };
+    default:
+      return state;
+  }
+};
+
+export interface FetchHookOptions extends RequestInit {
+  stream?: boolean;
+  json?: boolean;
+}
+
+const startRequestByDefault = (options?: FetchHookOptions) => !options || !options.method || options.method === "GET";
+
+/**
+ * React hook to make an HTTP request to a Gadget backend HTTP route. Preserves client side session information and ensures it's passed along to the backend.
+ *
+ * Returns a tuple with the current state of the request and a function to send or re-send the request. The state is an object with the following fields:
+ * - `data`: the response data, if the request was successful
+ * - `fetching`: a boolean describing if the fetch request is currently in progress
+ * - `error`: an error object if the request failed in any way
+ *
+ * The second return value is a function for executing the fetch request. It returns a promise for the response body.
+ *
+ * By default, `GET` requests are sent as soon as the hook executes. Any other request methods are not sent automatically, and must be triggered by calling the `execute` function returned in the second argument.
+ *
+ * Pass the `{ json: true }` option to expect a JSON response from the server, and to automatically parse the response as JSON. Otherwise, the response will be returned as a `string` object.
+ *
+ * Pass the `{ stream: true }` to get a `ReadableStream` object as a response from the server, allowing you to work with the response as it arrives. Otherwise, the response will be returned as a `string` object.
+ *
+ * If you want to read model data, see the `useFindMany` function and similar. If you want to invoke a backend Action, use the `useAction` hook instead.
+ *
+ * @param path the backend path to fetch
+ * @param options the `fetch` options for the request
+ *
+ * @example
+ * ```
+ * export function UserByEmail(props: { email: string }) {
+ *   const [{data, fetching, error}, refresh] = useFetch("/users/get", {
+ *     method: "GET",
+ *     body: JSON.stringify({ email: props.email }})
+ *     headers: {
+ *       "content-type": "application/json",
+ *     }
+ *     json: true,
+ *   });
+ *
+ *   if (result.error) return <>Error: {result.error.toString()}</>;
+ *   if (result.fetching && !result.data) return <>Fetching...</>;
+ *   if (!result.data) return <>No user found with id={props.id}</>;
+ *
+ *   return <div>{result.data.name}</div>;
+ * }
+ */
+export function useFetch(path: string, options: { stream: true } & FetchHookOptions): FetchHookResult<ReadableStream<Uint8Array>>;
```

Comment:
Curious: instead of building our own thing, do you think it would be better to build on top of something like https://use-http.com/?

Reply:
I would love to not have to maintain this ourselves, but I looked at a couple of the options and I found they were bulky in terms of bundle size, and had fanciful APIs that we don't really need. Didn't go into enough detail in the PR description but `use-http` returns this big object you can then use to call something with any verb and kind of use for both reads and writes to your backend via REST API -- I think that's just kind of overkill for us. We want to encourage use of the GraphQL API for everything you can use it for, and this is more for tiny little things that you can kind of forget about.

Reply:
I'm a-o-k going with what we have if you've dug in some. I think the main thing is if we're happy with the API we can always return and build on another library.

I think another option is to make it easy for others to use whatever fetch library they prefer, if they want to. To do that we could expose some of the stuff our own `connection.fetch` does, eh? No rush on that though.